src/weakptr.c

```diff
@@ -234,22 +235,16 @@ Int LengthWPObj(Obj wp)
     return len;
 #endif
 
-#ifndef USE_BOEHM_GC
```

Comment:
Why do this? The logic here has been changed quite confusingly. Before I try tracing it out, have you changed anything, or just moved things around?

Reply:
The code *was* changed in a non-trivial way, though it should be 100% identical in behavior to the original.

The reason I did this was to make the code easier to understand -- at least to me, it was very difficult to grok the original. Case in point: it turns out that the GASMAN and Boehm code are really doing the same thing, it just was difficult to see.

To elaborate, before the change, this code looked like this:
```C
#ifndef USE_BOEHM_GC
  Obj elm;
  while (len > 0 && 
         (!(elm = ELM_WPOBJ(wp,len)) ||
          IS_WEAK_DEAD_BAG(elm))) {
    changed = 1;
    if (elm)
      ELM_WPOBJ(wp,len) = 0;
    len--;
  }
#else
  while (len > 0 && !ELM_WPOBJ(wp, len)) {
    changed = 1;
    len--;
  }
#endif
```
Can you see that the Boehm and GASMAN code are doing exactly the same thing? I couldn't...

Now, I am already wary of simple assignments in loop conditionals, but if they look like `while ((elm = ELM_PLIST(list, n) { ...` I'd accept them, as it's reasonable easy to discern the meaning. But here the assignment was nested into a complex logical expression. Tracing out what it did was IMHO non-trivial. So I refactored it: Essentially, change the loop to `while (1)`, then move the condition into the loop body as `if (!cond) break;`, then apply boolean logic to transform and the split the condition... The result is the new loop:
```C
  while (len > 0) {
    elm = ELM_WPOBJ(wp, len);
    if (elm && !IS_WEAK_DEAD_BAG(elm))
        break;
    changed = 1;
    if (elm)
      ELM_WPOBJ(wp, len) = 0;
    len--;
  }
```
Now you can compare this to the Boehm GC code, and realize the following: there is an additional `IS_WEAK_DEAD_BAG`, but for Boehm GC, that does nothing. So we abort the loop via `break` as soon as `elm != 0`. But in that case, the second `if` never is reached; or in other words, whenever it is reached, we'll have `elm ==0`. So the Boehm version is really equivalent. And in fact, the compiler notices that and optimize the second `if` statement away.

Reply:
Slight correction: with Boehm GC, `IS_WEAK_DEAD_BAG` still does something, but it's this:
```C
#define IS_WEAK_DEAD_BAG(bag) (!(bag))
```
so `if (elm && !IS_WEAK_DEAD_BAG(elm))` becomes `if (elm && !!elm)` but again, the compiler is smart enough to figure out this is the same as `if (elm)`